PP-864: Ability to suspend jobs on Cray

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-864](https://pbspro.atlassian.net/browse/PP-864)**
* **[PP-948](https://pbspro.atlassian.net/browse/PP-948)**

#### Problem description
Support for suspend/resume of jobs on Cray X* series systems

#### Cause / Analysis
New feature for Cray X* series platform. 

#### Solution description
Changes are mostly related to switching IN/OUT ALPS reservations.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

####NOTE
- This change also consists of a bug fix for **[PP-948](https://pbspro.atlassian.net/browse/PP-948)** . Change is in node_manager.c where we were dereferencing an invalid memory location.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

lisa-altair commented:
One comment before I even start reviewing: PP-948 should also be listed in the Issue-ID section.

vccardenas commented on test/tests/functional/pbs_cray_suspend_resume.py:
```diff
@@ -0,0 +1,91 @@
+    """
+    def setUp(self):
+        if self.du.get_platform() is not "cray":
+            self.skipTest("Test suite only meant to run on a Cray")
```
You might also want to skip if suspend-resume is not enabled on the Cray.

arungrover replied:
@vccardenas Thanks for reviewing!
I can skip the test but just checking a parameter in a config file to decide that suspend/resume on the system is enabled would not be enough because for this configuration to take effect even ALPS needs to be restarted.
Secondly, there might be cases where ALPS config file may be present in different path in different CLEs.
For the reasons stated above I'd rather add a comment in the PTL script stating that the script expects suspend/resume being enabled on the system.

lisa-altair replied:
Yes, adding a comment seems like a good idea.

lisa-altair commented on src/resmom/requests.c:
```diff
@@ -1622,12 +1626,166 @@ post_resume(job *pjob, int err)
+		sprintf(log_buffer, "Timed out after %d attempts over %ld "
+			"seconds of attempting to confirm the ALPS "
+			"SWITCH completed.", i, total_time);
+		return (PBSE_ALPS_SWITCH_ERR);
```
Are we missing a log_event here?  Or will this log buffer get printed in the calling function?

arungrover replied:
Good catch, here and on many other places

lisa-altair replied:
I don't think that errors should be event type="PBSEVENT_DEBUG2" or log type of "LOG_DEBUG".   It should probably be PBSEVENT_SYSTEM and LOG_ERR.

arungrover replied:
Agreed taht DEBUG2 and LOG_DEBUG does not go well in this place but I don't think it is a system call failure to list it as PBSEVENT_SYSTEM. In reality the call did go through, it's just that the result turns out as EMPTY. I think it is better suited as PBSEVENT_DEBUG and LOG_ERR. Let me know what you think

lisa-altair replied:
Alright, I see your point.  Your suggestion of PBSEVENT_DEBUG and LOG_ERR work for me.  Thanks!

lisa-altair commented on src/resmom/requests.c:
```diff
@@ -1622,12 +1626,166 @@ post_resume(job *pjob, int err)
+		return (PBSE_NONE);
+	} else if (rc < 0) {
+		sprintf(log_buffer, "Fatal ALPS QUERY of status.");
+		return (PBSE_ALPS_SWITCH_ERR);
```
Do we need a call to log_event() here?

arungrover replied:
added

lisa-altair replied:
Since these are errors, you should use PBSEVENT_SYSTEM and LOG_ERR.

arungrover replied:
I'll change it

lisa-altair commented on src/resmom/requests.c:
```diff
@@ -1622,12 +1626,166 @@ post_resume(job *pjob, int err)
+		sprintf(log_buffer, "Timed out after %d attempts over "
+			"%ld seconds of waiting for a status of EMPTY "
+			"to change.  Proceeding as if the SWITCH "
+			"succeeded.", i, total_time);
```
If we proceed as if the SWITCH succeeded will PBS end up seeing problems when it tries to make a new ALPS reservation and ALPS doesn't allow it?

arungrover replied:
You are right. I should through a switch error instead

lisa-altair replied:
You changed the return value, but now it conflicts with the log message.

arungrover replied:
I'll remove this sentence

arungrover replied:
@lisa-altair I take the change back because for an empty reservation being reported by ALPS we should just continue

lisa-altair replied:
Oh my gosh, you are correct @arungrover.  For an EMPTY response we are good to continue.

mike0042 replied:
I'm fine with this change.

lisa-altair commented on src/resmom/requests.c:
```diff
@@ -1622,12 +1626,166 @@ post_resume(job *pjob, int err)
  * @param[in] pjob - pointer to job
  * @param[in] which - indication for whether suspend/resume
+ *		0 - resume
+ *		1 - suspend
```
Perhaps we should create 2 #define for these?  Then we can use them in the code instead of always having to check what does 0 mean?

arungrover replied:
Added

arungrover commented on src/resmom/linux/alps.c:
```diff
@@ -2985,101 +2991,236 @@ reservation_start(ud_t *d, const XML_Char *el, const XML_Char **atts)
+		 * syntax is just a fancy way of stepping through the pairs.
+		 */
+		for (np=vp=atts, vp++; np && *np && vp && *vp; np=++vp, vp++) {
+		xml_dbg("%s: %s = %s", __func__, *np, *vp);
```
Done

arungrover commented on src/resmom/linux/alps.c:
```diff
@@ -4301,6 +4457,9 @@ free_basil_response_data(basil_response_t *brp)
 			if (brp->data.query.data.engine.basil_support)
 				free(brp->data.query.data.engine.basil_support);
+		} else if (brp->data.query.type == basil_query_status) {
+			if (brp->data.query.data.status.reservation)
```
removed it

arungrover commented on src/resmom/linux/alps.c:
```diff
@@ -5683,6 +5855,301 @@ alps_cancel_reservation(job *pjob)
+		free_basil_response_data(brp);
+		return (2);
+	}
+	/* Getting the status of EMPTY means there was nothing to do, so
```
Changed

arungrover commented:
@lisa-altair Please review it again. I've addressed your comments.
@PBSPro/pbspro-maintainers  - there seems to be a problem with appveyor CI. It seems it is not able to build when there are conflicts in a pull. Having conflicts in the pull is pretty common because the code is every changing. Appveyor should just build the changes from the pushing branch only

vccardenas commented on test/tests/functional/pbs_cray_suspend_resume.py:
```diff
@@ -0,0 +1,91 @@
+        """
+        Check that on Cray restrict_res_to_release_on_suspend is always set
+        to 'ncpus' by default
+        """
```
@arungrover , shouldn't we test this also on Cray ALPS simulator due to changes in test/fw/ptl/lib/pbs_testlib.py, lines 4892 and 4893?

arungrover replied:
@vccardenas I did post test results from simulator on the ticket, I shall do that again and post it

vccardenas commented on test/tests/functional/pbs_cray_suspend_resume.py:
```diff
@@ -0,0 +1,91 @@
+            self.server.sigjob(jobid=jid, signal="suspend")
+        except PbsSignalError as e:
+            self.assertTrue("Switching ALPS reservation failed" in e.msg[0])
+
```
@arungrover , wouldn't you want to check for the new error code 15219 too?

arungrover replied:
PTL may not show error code in API mode that is why I did not check for that

vccardenas commented on test/tests/functional/pbs_cray_suspend_resume.py:
```diff
@@ -0,0 +1,93 @@
+        """
+
+        msg_expected = "BASIL;ERROR: ALPS error: apsched: \
+at least resid .* is exclusive"
```
Since the expected message is unstable per EDD perhaps you could you add a note in the test's docstring ?

arungrover replied:
I'll add it

vccardenas replied:
There is still a risk that the test will later fail if the expected message, which is unstable, changes.  Adding the note does not prevent this from happening. If it is important enough to test for this and that the test actually needs to check for the log message then perhaps the interface should be made stable?  Or perhaps there is another way of testing this functionality without having to check for the unstable log message?

bhroam replied:
I'm with Voltaire on this one.  If you are testing an interface in an automated test, it is a stable interface.  Unstable means the message can change at any time.  If that happens, your test would break.

lisa-altair commented:
Thanks for making the product code changes @arungrover.  My only remaining comments are for the new PTL file.

arungrover commented on test/tests/functional/pbs_cray_suspend_resume.py:
```diff
@@ -0,0 +1,93 @@
+        except PbsSignalError as e:
+            self.assertTrue("Switching ALPS reservation failed" in e.msg[0])
+
+        self.server.expect(JOB, 'exec_vnode', id=jid, op=SET)
```
you are right, It should be exec_host rather than vnode. I'll make the change

arungrover commented:
@lisa-altair, @vccardenas  please have a look again

lisa-altair commented on test/tests/functional/pbs_cray_suspend_resume.py:
```diff
@@ -67,6 +67,8 @@ def test_exclusive_job_not_suspended(self):
         """
         If a running job is a job with exclusive placement then this job can
         not be suspended.
+        This test is checking for a loge message which is an unstable
```
Minor comment: misspelled "log"

arungrover replied:
@vccardenas & @lisa-altair I've removed the comment and made the log interface stable.

vccardenas replied:
@arungrover , thanks for updating the EDD and the test (removed comment and expecting exec_host to be set).  Looks good to me now.

arungrover replied:
Ahh... It just never ends... :)
rectified in latest push

arungrover commented on test/tests/functional/pbs_cray_suspend_resume.py:
```diff
@@ -0,0 +1,93 @@
+    suspend/resume feature is enabled on it.
+    """
+    def setUp(self):
+        if not self.du.get_platform().startswith('cray'):
```
Sorry about a late reply
While this is true that platform types can have something else starting with 'cray' in them, I see the likelihood of this happening to be rare and even if it happens it will be very rare that the test will fail because it will be some kind of 'cray' platform.
I also feel that a change in get_platform() from PTL will be published and we can use that information to decide whether or not the test needs a change.
what do you think?

lisa-altair replied:
Thanks for the response.  That makes sense to me.  I am fine with it.

arungrover commented:
@lisa-altair @vccardenas Added back the comment about unstable interface in the ptl script. Please review it again

mike0042 commented on src/resmom/linux/alps.c:
```diff
@@ -4447,6 +4605,7 @@ alps_request_parent(int fdin)
 		free_basil_response_data(brp);
 		return (NULL);
 	}
+
```
I don't like blank lines. Didn't like them more than a decade ago, and I still don't like them.

arungrover replied:
I'll remove it

mike0042 commented on src/resmom/linux/alps.c:
```diff
@@ -5102,8 +5262,19 @@ alps_create_reserve_request(job *pjob, basil_request_reserve_t **req)
+			 * Set it in the ALPS reservation
+			 */
+			p->rsvn_mode = basil_rsvn_mode_exclusive;
+		}
```
What was wrong with the tertiary operator? The conditional has changed, but it saves space.

Why did the conditional change?

mike0042 commented:
You don't need to squash, but please rebase with your next commit to resolve conflicts.

lisa-altair commented on src/resmom/mom_main.c:
```diff
@@ -263,9 +263,16 @@ time_t		time_resc_updated = 0;
+int		alps_release_timeout = ALPS_REL_TIMEOUT;
+int		alps_confirm_empty_timeout = ALPS_CONF_EMPTY_TIMEOUT;
+int		alps_confirm_switch_timeout = ALPS_CONF_SWITCH_TIMEOUT;
+int		alps_create_vnode_per_numa = TRUE;
```
Good catch, there should not be alps_create_vnode_per_numa.

arungrover commented on src/resmom/mom_main.c:
```diff
@@ -4737,9 +4786,11 @@ read_config(char *file)
-	alps_release_timeout      = 600;	/* 10 min */
+	alps_release_timeout      = ALPS_REL_TIMEOUT;
+	alps_confirm_empty_timeout = ALPS_CONF_EMPTY_TIMEOUT;
+	alps_confirm_switch_timeout = ALPS_CONF_SWITCH_TIMEOUT;
```
You are right, this is not needed. I'll remove the first initialization

arungrover commented on src/resmom/requests.c:
```diff
@@ -1622,12 +1628,171 @@ post_resume(job *pjob, int err)
+	 * If not, just return PBSE_NONE.  
+	 */
+	if (pjob->ji_extended.ji_ext.ji_reservation <= 0) {
+		return PBSE_NONE;
```
I think this condition is added just for safety when a job may not report alps reservation. This to me is unlikely. I also think that switch error are more suited for cases where we tried a switch and failed, in this case we can not even try a switch because there is no reservation.
what do you think?

mike0042 replied:
I suppose it depends on how the caller(s) handle the return code. If they don't care whether the reservation exists, then this is as good a way as any to handle the situation. At the very least, we should be printing a log message when this condition occurs... probably an INFO or WARN.

lisa-altair replied:
This is handling the case where a job running on a login node needs to be suspended, in which case there is no need to talk to ALPS to do a SWITCH.

mike0042 replied:
Not exactly. This is handling a case where the reservation doesn't exist on the mom. Still the case that ALPS doesn't have to perform a switch, but mom expected the reservation to be there. We should log something when this happens.

arungrover replied:
I was wrong here, this code is not only for safety but there is another use case where it will come in handy - when the job runs on login node and that job is being preempted. This is an unlikely scenario but it can still happen. In those cases we might to just continue as we would on any non cray platform and then suspend the job. So, I'm also of the opinion to not log a message. What do you think @mike0042 ?

arungrover commented on src/resmom/requests.c:
```diff
@@ -1622,12 +1628,171 @@ post_resume(job *pjob, int err)
+	end_time = begin_time;
+	i = 0;
+	while ((total_time = end_time - begin_time) < timeout_val) {
+		i++;
```
done

arungrover commented on src/server/node_manager.c:
```diff
@@ -5714,7 +5714,7 @@ write_single_node_mom_attr(struct pbsnode *np)
 
 	attr.parent_id = np->nd_name;
-	attr.attr_resc[0] = 0;
+	attr.attr_resc = "";
```
It isn't allocated. This is just a pointer. This is what was causing server crash reported in PP-948

mike0042 replied:
So long as we never try to realloc attr_resc, this is fine.

arungrover commented on src/server/req_signal.c:
```diff
@@ -529,7 +529,12 @@ create_resreleased(job *pjob)
+	 * (node:resc=n)+(node:resc=m). This will add room for this
+	 * expansion.
+	 */
+	resreleased = (char *) calloc(1, strlen(pexech->at_val.at_str)*2 + 1);
```
It is really hard to get the exact value we will be needing, implementation is just taking the safest guess and using that for allocation.

mike0042 replied:
Not a problem. Was hoping there might be a better way.

arungrover commented:
@mike0042 I've rebased after the last commit and pushed the changes you suggested. Please have a look.

mike0042 commented on src/resmom/linux/alps.c:
```diff
@@ -1003,7 +1001,14 @@ response_data_start(ud_t *d, const XML_Char *el, const XML_Char **atts)
+				 * response. The other types can get set in
+				 * inventory_start and engine_start.
+				 */
+				brp->data.query.type = basil_query_status;
```
Hard to say how many times this code has changed over the years, but the original intent was to initialize the query type to none. It is later used to determine which (if any) memory needs to be freed in free_basil_resposne_data(). The handler for the engine query (engine_start) sets query type to basil_query_engine. I would expect the other handlers to do the same. Setting the query type to status could have unintended effect.

arungrover replied:
@mike0042 Since there is no other way for us to identify that the query type as status, this seems to be the only viable option for now. I'd request you to please consider it again

arungrover commented on src/cmds/scripts/pbs_habitat.in:
```diff
@@ -957,6 +957,11 @@ if [ "${PBS_START_SERVER:-0}" != 0 ] ; then
+			${PBS_EXEC}/bin/qmgr <<-EOF > /dev/null
+				set server restrict_res_to_release_on_suspend = ncpus
+			EOF
+		fi
```
there is a subtle difference in both the cases. In case of upgrade it appends ncpus and in case of install it sets it to only ncpus

arungrover commented on src/resmom/requests.c:
```diff
@@ -1622,12 +1628,171 @@ post_resume(job *pjob, int err)
+	 * as per the Cray ALPS folks, and we will poll for a successful 
+	 * suspend state here.  If the confirmation takes a while, then PBS
+	 * may need to somehow poll for the confirmation without tying up
+	 * the mom.
```
I'm not sure if async notification would report/confirm a switch event that is why I'm a little skeptical about adding such a comment.

lisa-altair commented on src/resmom/linux/alps.c:
```diff
@@ -799,12 +799,10 @@ stack_busted(ud_t *d)
+				if (strcmp(BASIL_ELM_RESPONSEDATA, prev) != 0) {
+					parse_err_illegal_start(d);
+					return (1);
+				}
```
This is wrong, we cannot check for the query type in stack_busted(), because brp->data.query.type doesn't get set until we start parsing the values within the XML.

arungrover replied:
That's correct Lisa, I don't know what I was thinking. You are right this will fail, I don't see any way to work around it because in stack busted we don't know which query type we are handling. I shall take this off.

mike0042 replied:
@lisa-altair are you certain that is the case? When I look at the code prior to these changes, I can clearly see a test being performed on data.query.type. Are you saying the existing code was incorrect?

lisa-altair replied:
@mike0042 you're right, it could be checked.  But wouldn't that be verifying the order in which the PBS code ran (because we expected basil_query_status to be set), as opposed to verifying the structure of the XML elements?

arungrover replied:
I see that a check for "data.query.type" is there at line 694 but I am confused as how it is useful.
we check for node array element to come after Inventory and if it is not inventory the we throw an error. But, if the check passes then we check for data.query.type to be anything other than inventory, which will be false no matter what, because "data.query.type" is set in inventory_start.

mike0042 replied:
That check was there for a reason. A reason that may no longer apply. I think my concern is based on some fuzzy memories of how the code used to work long ago. I don't see a reason for holding this back any longer, and I can't find a good reason to change what you currently have in the PR.

arungrover commented:
@mike0042 are you satisfied with the changes?

mike0042 commented:
Changes approved @arungrover.

arungrover commented:
@mike0042 shall I rebase and squash changes?

arungrover commented:
@mike0042 & @lisa-altair  while testing this change further I found a problem when ALPS reports switch out status as EMPTY. Ideally we should just be considering as switched and continue but I was considering that scenario as a switch error. I've pushed the fix in, please review it.

mike0042 commented:
@arungrover please rebase and squash.
